tests/kernel/device_new/src/main.c

```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright (c) 2022 Trackunit A/S
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/ztest.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/spi.h>
+
+#define DUMMY1_NODE_ID DT_NODELABEL(dummy1)
+#define DUMMY2_NODE_ID DT_NODELABEL(dummy2)
+#define DUMMY_LOW_NODE_ID DT_NODELABEL(dummy_low)
+#define DUMMY_HIGH_NODE_ID DT_NODELABEL(dummy_high)
+
+/* APIs implemented by dummy driver defined in dummy_driver.c */
+extern const struct uart_driver_api dummy_uart_api;
+extern const struct i2c_driver_api dummy_i2c_api;
+extern const struct spi_driver_api dummy_spi_api;
+
+/* Run counter used to verify dummy driver init has been invoked correctly */
+extern int dummy_driver_init_run_cnt;
+
+/*
+ * Catch any linker errors here. If external references are generated and
+ * included, and the driver is able to instanciate them, this will not cause
+ * any errors.
+ */
+static const struct device *uart_dev1 = DEVICE_DT_API_GET(DUMMY1_NODE_ID, uart);
+static const struct device *i2c_dev1 = DEVICE_DT_API_GET(DUMMY1_NODE_ID, i2c);
+static const struct device *spi_dev1 = DEVICE_DT_API_GET(DUMMY1_NODE_ID, spi);
+
+static const struct device *uart_dev2 = DEVICE_DT_API_GET(DUMMY2_NODE_ID, uart);
+static const struct device *i2c_dev2 = DEVICE_DT_API_GET(DUMMY2_NODE_ID, i2c);
+static const struct device *spi_dev2 = DEVICE_DT_API_GET(DUMMY2_NODE_ID, spi);
+
+/* Test DEVICE_DT_API_SUPPORTED macro */
+#if (DEVICE_DT_API_SUPPORTED(DUMMY1_NODE_ID, uart) == 0)
+	#warning "DEVICE_DT_API_SUPPORTED failed for dummy1 uart API"
+#endif
+
+#if (DEVICE_DT_API_SUPPORTED(DUMMY1_NODE_ID, i2c) == 0)
+	#warning "DEVICE_DT_API_SUPPORTED failed for dummy1 i2c API"
+#endif
+
+#if (DEVICE_DT_API_SUPPORTED(DUMMY1_NODE_ID, spi) == 0)
+	#warning "DEVICE_DT_API_SUPPORTED failed for dummy1 spi API"
+#endif
+
+#if (DEVICE_DT_API_SUPPORTED(DUMMY1_NODE_ID, led) == 1)
+	#warning "DEVICE_DT_API_SUPPORTED failed for dummy1 led API"
+#endif
+
+#if (DEVICE_DT_API_SUPPORTED_ANY(uart) == 0)
+	#warning "DEVICE_DT_API_SUPPORTED_ANY failed for uart API"
+#endif
+
+#if (DEVICE_DT_API_SUPPORTED_ANY(dummy) == 1)
+	#warning "DEVICE_DT_API_SUPPORTED_ANY failed for dummy API"
+#endif
```

Comment:
build asserts?

Reply:
Yes, i will change them to build asserts